src/analysis/souffle/tags.dl

```diff
@@ -28,4 +28,21 @@
 .type Tag <: symbol
 .decl isTag(tag: Tag)
 
+// An access path marked with an integrity tag (shortened here to IntegTag)
+// indicates that some property must be true of the data at the given
+// AccessPath. The conservative over-approximation for propagating an integrity
+// tag is that a node has an integrity tag only if all of its predecessors have
+// that integrity tag. Checking this property requires an intersection, which
+// unfortunately cannot be directly expressed in datalog.
+.type IntegTag <: symbol
+
+// To deal with that, we instead propagate the opposite of the integrity tag:
+// the absence of an integrity tag, which we call an inverse integrity tag, or
+// an InvIntegTag. An inverse integrity tag is introduced any node with no
+// predecessor edges and without an integrity tag added. It is also introduced
+// at any point where an integrity tag is removed. It is removed at any point
+// where an integrity tag is added. Textually, an InvIntegTag symbol is just the
+// associated IntegTag with a '!' prepended.
+.type InvIntegTag <: symbol
```

Comment:
Why do we need this? Can't we capture this in a relation like `lacksIntegrity(ap: AccessPath, owner: Principal, tag: IntegrityTag)`?

Reply:
That is also fine, and semantically equivalent to what I've done with the inverse integrity tags here. The reason I made the inverse integrity tags was to hopefully prevent two different types of confusion:

1. Improve readability by making the relations not contain negatives (that is, speaking about the positive presence of an inverse integrity tag rather than the absence of an integrity tag)

2. Use a separate base type to prevent propagating an integrity tag from a place where it was supposed to be absent to a place where it was supposed to be present.

That said, if this is just confusing in a different way, happy to just use `lacksIntegrity`.

Reply:
I've moved to using `lacksIntegrity` as suggested.